File: TeledonSockets/TeledonPersistence/src/main/java/repository/CharitableCaseRepository.java
package repository;

import model.CharitableCase;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class CharitableCaseRepository implements ICharitableCaseRepository {
    private JdbcUtils dbUtils;

    private static final Logger logger = LogManager.getLogger();

    public CharitableCaseRepository(Properties props) {
        logger.info("Initializing CharitableCaseRepository with properties: {}",props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public void save(CharitableCase elem) {
        logger.traceEntry("saving charitable case {}", elem);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into CharitableCases(name,amountRaised) values (?,?)")){
            preparedStatement.setString(1,elem.getName());
            preparedStatement.setInt(2,elem.getAmountRaised());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances",result);
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit();

    }

    @Override
    public void delete(Integer id) {
        logger.traceEntry("deleting charitable case {}", id);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from CharitableCases where id = ?")){
            preparedStatement.setInt(1,id);
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances",result);
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit();

    }

    @Override
    public void update(CharitableCase elem, Integer id) {
        logger.traceEntry("updating charitable case {}", id);
        Connection connection = dbUtils.getConnection();
        String sql = String.format("update CharitableCases set amountRaised = ? where id = %d",id);
        try(PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1,elem.getAmountRaised());
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit();


    }

    @Override
    public CharitableCase findById(Integer id) {
        logger.traceEntry("finding charitable case with id {}", id);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from CharitableCases where id = ?")){
            preparedStatement.setInt(0,id);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    int idCharitableCase = resultSet.getInt("id");
                    String nameCharitableCase = resultSet.getString("name");
                    int amountCharitableCase = resultSet.getInt("amountRaised");
                    CharitableCase charitableCase = new CharitableCase(nameCharitableCase,amountCharitableCase);
                    logger.traceExit(charitableCase);
                    return charitableCase;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit("No charitable case found with id {}", id);
        return null;

    }

    @Override
    public Iterable<CharitableCase> findAll() {
        logger.traceEntry("finding all charitable cases {}");
        Connection connection = dbUtils.getConnection();
        List<CharitableCase> charitableCaseList = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from CharitableCases")) {
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while(resultSet.next()){
                    int idCharitableCase = resultSet.getInt("id");
                    String nameCharitableCase = resultSet.getString("name");
                    int amountCharitableCase = resultSet.getInt("amountRaised");
                    CharitableCase charitableCase = new CharitableCase(nameCharitableCase,amountCharitableCase);
                    charitableCaseList.add(charitableCase);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit(charitableCaseList);
        return charitableCaseList;
    }

    @Override
    public int size() {
       logger.traceEntry();
       Connection connection = dbUtils.getConnection();
       try(PreparedStatement preparedStatement = connection.prepareStatement("select count(*) as [SIZE] from CharitableCases")){
           try(ResultSet resultSet = preparedStatement.executeQuery()){
               if(resultSet.next()){
                   logger.traceExit(resultSet.getInt("SIZE"));
                   return resultSet.getInt("SIZE");
               }
           }
       } catch (SQLException e) {
           logger.error(e);
           System.out.println("Error DB " + e);
       }
       return 0;
    }

    @Override
    public Integer getCharitableCaseByName(String name) {
        logger.traceEntry("finding charitable case with id {}");
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from CharitableCases where name = ?")){
            preparedStatement.setString(1,name);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    int idCharitableCase = resultSet.getInt("id");
                    return idCharitableCase;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB " + e);
        }
        logger.traceExit("No charitable case found with id {}");
        return null;

    }
}

File: teledonSocketsCs/persistence/CharitableCaseRepository.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Configuration;
using System.Data;
using System.Text;
using log4net;
using teledonCS.model;

namespace teledonCS.repository
{
    public class CharitableCaseRepository : ICharitableChaseRepository
    {
        private static readonly ILog log = LogManager.GetLogger("CharitableCaseRepository");
        IDictionary<String, string> props;

        public CharitableCaseRepository(IDictionary<String, string> props)
        {
            log.Info("Creating CharitableCaseRepository");
            this.props = props;
        }

        public void delete(int key)
        {
            log.InfoFormat("Deleting charitable case");
            IDbConnection con = DBUtils.getConnection();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from CharitableCases where id = @key";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@key";
                paramId.Value = key;
                comm.Parameters.Add(paramId);
                var dataR = comm.ExecuteNonQuery();
                
                if (dataR == 0)
                {
                    log.Error("No charitable case removed");
                }
                else
                    log.InfoFormat("Charitable case removed");
                    
            }
            
        log.InfoFormat("Exiting delete");
        }

        public List<CharitableCase> findAll()
        {
            log.InfoFormat("Finding all charitable cases");
            IDbConnection con = DBUtils.getConnection();
            IList<CharitableCase> charitableCases = new List<CharitableCase>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,name,amountRaised from CharitableCases";
                using (var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idCC = dataR.GetInt32(0);
                        String nameCC = dataR.GetString(1);
                        int amountRaisedCC = dataR.GetInt32(2);
                        CharitableCase charitableCase = new CharitableCase(idCC, nameCC, amountRaisedCC);
                        charitableCases.Add(charitableCase);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return (List<CharitableCase>) charitableCases;
        }

        public int? getCharitableCaseByName(string name)
        {
            log.InfoFormat("Entering find charitable case by name");
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id from CharitableCases where name = @nameToSearch";
                IDbDataParameter paramName = comm.CreateParameter();
                paramName.ParameterName = "@nameToSearch";
                paramName.Value = name;
                comm.Parameters.Add(paramName);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idCC = dataR.GetInt32(0);

                        log.InfoFormat("Exiting find by name");
                        
                        return idCC;
                    }
                }
            }

            log.InfoFormat("Exiting findOne with value{0}", null);
            return null;
            
        }

        public CharitableCase findOne(int key)
        {
            log.InfoFormat("Entering findOne with value{0}",key);
            IDbConnection con = DBUtils.getConnection();

            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,name,amountRaised from CharitableCases where id = @key";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@key";
                paramId.Value = key;
                comm.Parameters.Add(paramId);

                using (var dataR = comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idCC = dataR.GetInt32(0);
                        String nameCC = dataR.GetString(1);
                        int amountCC = dataR.GetInt32(2);
                        CharitableCase charitableCase = new CharitableCase(nameCC, amountCC);
                        charitableCase.id = idCC;
                        log.InfoFormat("Exiting findOne with value{0}",charitableCase);
                        return charitableCase;
                    }
                }
            }

            log.InfoFormat("Exiting findOne with value{0}", null);
            return null;
        }

        public void save(CharitableCase el)
        {
            log.InfoFormat("Saving charitable case");
            var con = DBUtils.getConnection();
            using(var comm=con.CreateCommand())
            {
                comm.CommandText = "insert into CharitableCases(name,amountRaised) values (@name,@amount)";
                var paramName = comm.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = el.name;
                comm.Parameters.Add(paramName);
                
                var paramAmount = comm.CreateParameter();
                paramAmount.ParameterName = "@amount";
                paramAmount.Value = el.amountRaised;
                comm.Parameters.Add(paramAmount);
                
                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    log.Error("No charitable case added");
                    throw new Exception("No charitable case added!!");
                }
                else
                {
                    log.InfoFormat("Charitable Case saved");
                }
            }
            log.InfoFormat("Exiting save");
        }

        public void update(CharitableCase el)
        {
            log.InfoFormat("Updating charitable case");
            var con = DBUtils.getConnection();
            using(var comm=con.CreateCommand())
            {
                comm.CommandText = "update CharitableCases set amountRaised = @amountNew where id = @id";
                var paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = el.id;
                comm.Parameters.Add(paramId);

                var paramAmount = comm.CreateParameter();
                paramAmount.ParameterName = "@amountNew";
                paramAmount.Value = el.amountRaised;
                comm.Parameters.Add(paramAmount);

                try
                {
                    var result = comm.ExecuteNonQuery();
                    if (result == 0)
                    {
                        log.Error("No charitable case updated");
                        throw new Exception("No charitable case updated!!");
                    }

                }
                catch (Exception e)
                {
                    Console.Error.WriteLine(e.Message);
                }
                
            }
           
            log.InfoFormat("Exiting update");
            }
        
    }
}
